go-controller/hybrid-overlay/pkg/controller/node.go

```diff
@@ -158,45 +159,59 @@ func (n *Node) Run(stopCh <-chan struct{}) error {
 
 // getNodeSubnetAndIP returns the node's hybrid overlay subnet and the node's
 // first InternalIP, or nil if the subnet or node IP is invalid
-func getNodeSubnetAndIP(node *kapi.Node) (*net.IPNet, net.IP) {
-	var cidr *net.IPNet
+func getNodeSubnetAndIP(node *kapi.Node) ([]*net.IPNet, []net.IP) {
+	var ips []net.IP
 
 	// Parse Linux node OVN hostsubnet annotation first
 	cidrs, _ := util.ParseNodeHostSubnetAnnotation(node)
-	if cidrs != nil {
-		// FIXME DUAL-STACK
-		cidr = cidrs[0]
-	} else {
+	if cidrs == nil {
 		// Otherwise parse the hybrid overlay node subnet annotation
+		// TODO : DUAL-STACK change the code when HO annotations for multiple subnets are in
 		subnet, ok := node.Annotations[types.HybridOverlayNodeSubnet]
 		if !ok {
-
 			klog.V(5).Infof("Missing node %q node subnet annotation", node.Name)
 			return nil, nil
 		}
-		var err error
-		_, cidr, err = net.ParseCIDR(subnet)
+		_, cidr, err := net.ParseCIDR(subnet)
 		if err != nil {
 			klog.Errorf("Error parsing node %q subnet %q: %v", node.Name, subnet, err)
 			return nil, nil
 		}
+		cidrs = append(cidrs, cidr)
 	}
 
-	nodeIP, err := houtil.GetNodeInternalIP(node)
+	nodeIPv4, nodeIPv6, err := houtil.GetNodeInternalIPs(node)
 	if err != nil {
 		klog.Errorf("Error getting node %q internal IP: %v", node.Name, err)
 		return nil, nil
 	}
-
-	return cidr, net.ParseIP(nodeIP)
+	// Return an error if any of the internal addresses are missing in case of dual stack
+	if len(cidrs) == 2 && (nodeIPv4 == "" || nodeIPv6 == "") {
+		klog.Errorf("Error getting node %q internal IP: %v", node.Name, err)
+		return nil, nil
+	}
+	// In case of dual stack make sure IPv4 subnet is the first element
+	if len(cidrs) == 2 && utilnet.IsIPv6CIDR(cidrs[0]) {
+		tmp := cidrs[0]
+		cidrs[0] = cidrs[1]
+		cidrs[1] = tmp
+	}
+	ips = make([]net.IP, 0)
+	if nodeIPv4 != "" {
+		ips = append(ips, net.ParseIP(nodeIPv4))
+	}
+	if nodeIPv6 != "" {
+		ips = append(ips, net.ParseIP(nodeIPv6))
+	}
+	return cidrs, ips
```

Comment:
We don't guarantee ordering of IPs anywhere else. You should just have `houtil.GetNodeInternalIPs` return an array of IPs, and not assume anything about the order.

Reply:
Done. I had done it this way because it simplified the code in rest of the places. But I get your point and made changes accordingly.